pipeline/trusted-resources/cmd/sign/main.go

```diff
@@ -71,36 +73,49 @@ func main() {
 	}
 	defer f.Close()
 
-	tsBuf, err := ioutil.ReadFile(*taskFile)
+	tsBuf, err := ioutil.ReadFile(*resourceFile)
 	if err != nil {
-		log.Fatalf("error reading task: %v", err)
+		log.Fatalf("error reading file: %v", err)
 	}
 
-	ts := &v1beta1.Task{}
-	if err := yaml.Unmarshal(tsBuf, &ts); err != nil {
-		log.Fatalf("error unmarshalling taskrun: %v", err)
+	var crd resourcesemantics.GenericCRD
+	switch *kind {
+	case "Task":
+		crd = &v1beta1.Task{}
+	case "Pipeline":
+		crd = &v1beta1.Pipeline{}
+	}
+	if err := yaml.Unmarshal(tsBuf, &crd); err != nil {
+		log.Fatalf("error unmarshalling Task/Pipeline: %v", err)
```

Comment:
Thanks! 😄 

This reminds me of the mutating webhook issue, I wonder what if we call `crd.SetDefaults(ctx)`, would it work to avoid possible issues for mutating webhooks? 

One reason the mutating wh will change the values is because tekton needs the resource to have default values for some fields but the users may forgot to configure them. So if before we sign it we call the SetDefaults directly, will that address this issue?

@wlynch what do you think?

Reply:
Haven't applied the suggested changes yet but would like some feedback on this before that. 😄

Reply:
To add more context: Yongxuan and I discussed about this topic a littble bit, our thought is we define a new function called `FormatTask|Pipeline` to be called before customers sign their task/pipeline. In that function, we format the task/pipeline inputs from customers. For example, set default values before signing. 

Standardizing customer inputs before signing sounds pretty reasonable. Meanwhile, most importantly, we don't need to worry about what if Pipeline Mutation Webhooks modify a signed Task/Pipeline, which in the end is an issue about Mutation Webhook ordering.

Reply:
Hmmmm - not sure!

One thing I'd be worried about is the crd.SetDefaults probably wouldn't be sufficient to capture everything - the Tekton Pipeline webhook happens to correspond to the Defaultable type, but there might be other webhooks that do not conform to that at all. We probably want to sign either the initial unmodified input or the final fully resolved object - if we can't guarantee the fully resolved object my gut feeling would be to stick with the original object? wdyt?

Reply:
yes that's one issue, there is the possibility that other mutating wh change the tasks. But if the users don't apply new mutating wh, then it will help to complete the yaml files I guess? I feel like it is a partial fix solution to this issue and better than nothing? What will be the cons of it?

So I think for current solution, we can document this can also inform the users of possible issues:
1) the non-standard usage of task or pipeline yaml files. The tekton mutating wh will populate default values for them and thus fail the verification. -> this may be fixed if we call SetDefaults before signing or either logging it to users
2) if the user deploy other mutating wh to change the task or pipeline, then this solution may fail. -> We have no idea what wh user will deploy, so maybe just document this. (it is very hard to meet all users requirements I guess T_T)

Reply:
Exactly, we did touch the topic you mentioned. The main concern would be what exactly Pipeline webhooks modify.
1.@yongxuangzhang talked to Christie about this issue weeks ago, Christie's suggestion was we can avoid modifying stuff in webhooks. Maybe we need to bring this up in the community.
2.@yongxuanzhang also found that the SetDefault modification is inevasible, so we start to think what's the best solution.

Verification webhooks are invoked after mutation webhooks, so there's not an easy way to verify signatures of Task/Pipeline API objects that are created through K8s API Server. Wdyt?

Reply:
I added a flag and users can choose whether use it or not. 😄

Reply:
sgtm. let's hold off on that for another PR though? (or at the very least update this PR to include the changes?)

Reply:
Sure! I changed the commit messages and pr description